diff -Naur tmp/Python-2.7.13/Objects/object.c Python-2.7.13/Objects/object.c
--- tmp/Python-2.7.13/Objects/object.c	2016-12-17 21:05:07.000000000 +0100
+++ Python-2.7.13/Objects/object.c	2018-12-09 15:27:27.000000000 +0100
@@ -29,8 +29,8 @@
 }
 #endif /* Py_REF_DEBUG */
 
-int Py_DivisionWarningFlag;
-int Py_Py3kWarningFlag;
+__thread int Py_DivisionWarningFlag;
+__thread int Py_Py3kWarningFlag;
 
 /* Object allocation routines used by NEWOBJ and NEWVAROBJ macros.
    These are used by the individual routines for object creation.
